azurerm/resource_arm_iothub_shared_access_policy.go

```diff
@@ -0,0 +1,354 @@
+package azurerm
+
+import (
+	"fmt"
+	"log"
+	"regexp"
+	"strings"
+
+	"github.com/Azure/azure-sdk-for-go/services/preview/iothub/mgmt/2018-12-01-preview/devices"
+	"github.com/hashicorp/go-multierror"
+	"github.com/hashicorp/terraform/helper/schema"
+	"github.com/hashicorp/terraform/helper/validation"
+	"github.com/terraform-providers/terraform-provider-azurerm/azurerm/helpers/suppress"
+	"github.com/terraform-providers/terraform-provider-azurerm/azurerm/helpers/tf"
+	"github.com/terraform-providers/terraform-provider-azurerm/azurerm/helpers/validate"
+	"github.com/terraform-providers/terraform-provider-azurerm/azurerm/utils"
+)
+
+func resourceArmIotHubSharedAccessPolicy() *schema.Resource {
+	return &schema.Resource{
+		Create: resourceArmIotHubSharedAccessPolicyCreateUpdate,
+		Read:   resourceArmIotHubSharedAccessPolicyRead,
+		Update: resourceArmIotHubSharedAccessPolicyCreateUpdate,
+		Delete: resourceArmIotHubSharedAccessPolicyDelete,
+		Importer: &schema.ResourceImporter{
+			State: schema.ImportStatePassthrough,
+		},
+
+		Schema: map[string]*schema.Schema{
+			"name": {
+				Type:             schema.TypeString,
+				Required:         true,
+				ForceNew:         true,
+				DiffSuppressFunc: suppress.CaseDifference,
+				ValidateFunc: validation.StringMatch(regexp.MustCompile(`[a-zA-Z0-9!._-]{1,64}`), ""+
+					"The shared access policy key name must not be empty, and must not exceed 64 characters in length.  The shared access policy key name can only contain alphanumeric characters, exclamation marks, periods, underscores and hyphens."),
+			},
+
+			"resource_group_name": resourceGroupNameSchema(),
+
+			"iothub_name": {
+				Type:             schema.TypeString,
+				Required:         true,
+				ForceNew:         true,
+				DiffSuppressFunc: suppress.CaseDifference,
```

Comment:
What is the reasoning behind suppressing the case here?

Reply:
IOTHub name is case-insensitive, that's why the case is suppressed here. I just looked at the azurerm_iothub resource and saw that the case is not suppresed there. This may be dangerous, as simply changing the case of the name, which should not have any effect, as it is still the same resource, will recreate the IoTHub and all the data stored in it will be lost :( 
However, I can remove the suppress func if you have some kind of convention to always have the IDs case-sensitive.

Reply:
@maxbog, while you are correct that this api seems to be case insensitive, others are not and we are trying to be more consistent across the provider. Given the existing resources are and a plan will notify the user of deletion if they change the case we would like the supress removed. 

Additionally, if casing is "iothub" and a user wishes it to be "IOThub" the supress will prevent this.

Thanks!

Reply:
OK then, I removed the suppress.

Reply:
Thank you 🙏